### PR TITLE
fix(redis): add startup retry backoff

### LINKED ISSUE #14: Add retry-with-backoff for Redis connection on startup
## User Request

The `notifications` service crashes immediately on startup if Redis isn't ready. In CI (k3d cluster), both the notifications pod and the `notifications-redis-master` StatefulSet start within ~1s of each other, but Redis takes ~15-20s to be ready. The notifications service does a single `Ping` with 5s timeout, fails, and calls `os.Exit(1)`.

This causes `CrashLoopBackOff` in CI and delays cluster readiness.

### Root Cause

In `cmd/notifications/main.go`, `ensureRedis()` has zero retry logic:

```go
func ensureRedis(ctx context.Context, client *redis.Client) error {
    ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
    defer cancel()
    if err := client.Ping(ctx).Err(); err != nil {
        return fmt.Errorf("redis ping: %w", err)
    }
    return nil
}
```

If `Ping` fails → `run()` returns → `main()` exits with code 1.

Similarly, `subscriber.Start()` in `internal/redis/subscriber.go` calls `pubsub.Receive(ctx)` with no retry.

### Required Changes

1. **`cmd/notifications/main.go`** — Replace single-shot `ensureRedis` with retry loop using exponential backoff:
   - Max 15 attempts
   - Base delay 500ms, max delay 5s
   - 3s timeout per ping attempt
   - Log each retry attempt at warn level
   - Total wait budget: ~45s (more than enough for Redis startup)

2. **`internal/redis/subscriber.go`** — Add similar retry for `pubsub.Receive(ctx)` in `Start()`, or rely on `ensureRedis` succeeding first (if Ping passes, Subscribe+Receive will also succeed).

3. **Release as patch version** (0.2.2 or next appropriate version).

### Validation

- `go build ./...`
- `go test ./...`
- Verify the retry loop by starting the service before Redis is available and confirming it waits and connects successfully.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- add exponential backoff retries for Redis ping on startup
- add retry with backoff for Redis subscriber start
- tighten subscriber start test timeout to avoid long retry waits

## Testing
- go build ./...
- go test ./...
- go vet ./...

Closes #14

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- go build ./...
- go test ./...
  - passed: 3, failed: 0, skipped: 0
- go vet ./...
  - lint: no issues

File: cmd/notifications/main.go [hunk 1]
```diff
@@ -135,11 +142,58 @@ func run() error {
...
+		if backoff > redisRetryMaxBackoff {
+			backoff = redisRetryMaxBackoff
+		}
 	}
-	return nil
+
+	if lastErr == nil {
+		return fmt.Errorf("redis ping: unknown error")
+	}
+	return fmt.Errorf("redis ping: %w", lastErr)
```
Comment (noa-lucent) [inline cmd/notifications/main.go]: [major] **Unreachable "just-in-case" fallback masking an impossible state.**

`lastErr` can never be `nil` here. The loop always either returns `nil` on success (line 156) or assigns `lastErr = err` (line 158) before reaching this point. If `redisRetryMaxAttempts >= 1`, this branch is dead code.

Returning a synthetic `"redis ping: unknown error"` silently hides what would be a logic bug. If this state were ever reached, it should fail loudly.

Replace with a panic/assertion:

```go
if lastErr == nil {
    panic("ensureRedis: exhausted retries with no recorded error")
}
return fmt.Errorf("redis ping: %w", lastErr)
```

The same issue applies to `subscribeWithRetry` in `subscriber.go` (line 184).

File: cmd/notifications/main.go [hunk 2]
```diff
@@ -135,11 +142,58 @@ func run() error {
...
+			timer.Stop()
+			return ctx.Err()
+		case <-timer.C:
+		}
+
+		backoff *= 2
+		if backoff > redisRetryMaxBackoff {
+			backoff = redisRetryMaxBackoff
+		}
 	}
```
Comment (noa-lucent) [inline cmd/notifications/main.go:172]: [major] **Duplicate retry logic — extract a shared helper.**

The retry-with-backoff loop in `ensureRedis` and `subscribeWithRetry` (in `subscriber.go`) are structurally identical:

1. Loop up to N attempts
2. Check `ctx.Err()`
3. Execute operation
4. On failure: log, sleep with exponential backoff clamped to max, select on ctx/timer
5. After exhaustion: return `lastErr`

The only difference is the operation itself (ping vs. subscribe+receive) and the logger access pattern. This is a textbook DRY violation that will drift over time (e.g., one gets a bug fix and the other doesn't).

Extract a generic retry helper, e.g.:

```go
// internal/retry/retry.go
func WithBackoff(ctx context.Context, cfg Config, op func(ctx context.Context) error) error
```

Then both call sites become one-liners. The retry constants can also be unified (they're currently identical but declared in two separate files).

File: cmd/notifications/main.go [hunk 3]
```diff
@@ -135,11 +142,58 @@ func run() error {
...
+		if attempt == redisRetryMaxAttempts {
+			break
+		}
+
+		delay := backoff
+		if delay > redisRetryMaxBackoff {
+			delay = redisRetryMaxBackoff
+		}
+
+		if logger != nil {
```
Comment (noa-lucent) [inline cmd/notifications/main.go]: [minor] **Defensive nil check on `logger` is unnecessary at this call site.**

`ensureRedis` is called exactly once, from `run()`, where `logger` is always initialized (line 49). This `if logger != nil` guard is "just-in-case" defensiveness.

If you extract a shared retry helper (see other comment), the nil-safety concern moves to the helper's API. Otherwise, either require a non-nil logger in the function contract or use `zap.NewNop()` at the call site to guarantee it.

File: cmd/notifications/main.go [hunk 4]
```diff
@@ -135,11 +142,58 @@ func run() error {
...
+			return ctx.Err()
+		}
+		if attempt == redisRetryMaxAttempts {
+			break
+		}
+
+		delay := backoff
+		if delay > redisRetryMaxBackoff {
+			delay = redisRetryMaxBackoff
+		}
```
Comment (noa-lucent) [inline cmd/notifications/main.go]: [nit] **Redundant delay clamping.**

`delay` is clamped to `redisRetryMaxBackoff` here (lines 167-169), but `backoff` itself is already clamped at the bottom of the loop (lines 189-191). On the first iteration `backoff == redisRetryInitialBackoff` (500ms) which is already < `redisRetryMaxBackoff` (5s), and on all subsequent iterations `backoff` was clamped at the end of the previous iteration.

The first clamp is dead code. You can remove it and just use `backoff` directly as the delay:

```go
delay := backoff
```

Same applies to `subscribeWithRetry`.

File: internal/redis/subscriber.go [hunk 5]
```diff
@@ -125,6 +128,65 @@ func (s *Subscriber) Start(ctx context.Context) error {
...
+		backoff *= 2
+		if backoff > subscribeRetryMaxBackoff {
+			backoff = subscribeRetryMaxBackoff
+		}
+	}
+
+	if lastErr == nil {
+		return nil, fmt.Errorf("redis subscribe failed")
+	}
+	return nil, lastErr
```
Comment (noa-lucent) [inline internal/redis/subscriber.go]: [major] **Same unreachable "just-in-case" fallback as in `ensureRedis`.**

`lastErr` is always set before the loop can exit without a successful return. The `lastErr == nil` branch (line 184) is dead code. Replace with a panic/assertion to surface logic bugs rather than returning a synthetic error string.

```go
if lastErr == nil {
    panic("subscribeWithRetry: exhausted retries with no recorded error")
}
return nil, lastErr
```

File: internal/redis/subscriber_test.go [hunk 6]
```diff
@@ -66,7 +66,8 @@ func TestSubscriber_StartAndReceive(t *testing.T) {
 func TestSubscriber_Start_ErrorOnSubscribe(t *testing.T) {
 	t.Parallel()
 
-	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
```
Comment (noa-lucent) [inline internal/redis/subscriber_test.go:69]: [minor] **Test relies on context timeout to end retries, but doesn't verify retry behavior.**

The 50ms timeout is a good pragmatic fix to prevent the test from hanging, but the test only asserts that `Start` returns an error — it doesn't verify that any retry actually happened (e.g., that `Subscribe` was called more than once, or that the warn log was emitted).

Consider adding an attempt counter to `mockSubscriber` and asserting `mock.calls > 1` to confirm retries are exercised. This would catch regressions where the retry loop is accidentally removed or short-circuited.

File: cmd/notifications/main.go [hunk 7]
```diff
@@ -135,11 +142,58 @@ func run() error {
 	return nil
 }
 
-func ensureRedis(ctx context.Context, client *redis.Client) error {
-	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
-	defer cancel()
-	if err := client.Ping(ctx).Err(); err != nil {
-		return fmt.Errorf("redis ping: %w", err)
+func ensureRedis(ctx context.Context, logger *zap.Logger, client *redis.Client) error {
```
Comment (noa-lucent) [inline cmd/notifications/main.go:146]: [minor] **No unit tests for `ensureRedis` retry logic.**

`ensureRedis` now contains non-trivial retry logic (backoff progression, context cancellation during sleep, max-attempt exhaustion) but has zero test coverage. The subscriber side gets minimal indirect coverage via `TestSubscriber_Start_ErrorOnSubscribe`, but `ensureRedis` in `cmd/notifications/main.go` has no tests at all.

If you extract a shared retry helper (see other comment), a single well-tested helper would cover both paths. Otherwise, consider adding tests here — at minimum: success on first try, success after N failures, exhaustion, and context cancellation mid-retry.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for adding the retry logic — the overall approach is sound and addresses the issue well.

Two major items need attention before merge:

1. **Duplicate retry logic**: `ensureRedis` and `subscribeWithRetry` are near-identical 40-line retry loops. Extract a shared `retry.WithBackoff` helper to eliminate the duplication and unify the retry constants.

2. **Unreachable "just-in-case" fallbacks**: Both functions have a `lastErr == nil` guard that can never trigger. These should be panics (fail-loud) rather than synthetic error strings that would hide a logic bug.

Additional minor items: unnecessary `logger != nil` guard, redundant delay clamping, and missing test coverage for the retry logic (especially `ensureRedis` which has zero tests).

**Comment by casey-brooks** [comment]
## Summary
- added shared retry helper with backoff configuration in `internal/retry`
- refactored Redis startup ping and subscribe retries to use the helper
- added unit tests for retry behavior and context handling

## Testing
- go build ./...
- go test ./...
- go vet ./...

**Comment by casey-brooks** [comment]
## Summary
- added shared retry helper with backoff configuration in `internal/retry`
- refactored Redis ping and subscribe retries to use the helper
- added unit tests covering retry behavior and context cancellation

## Testing
- go build ./...
- go test ./...
  - passed: 4, failed: 0, skipped: 0
- go vet ./...
  - lint: no issues

File: internal/retry/retry.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+		}
+
+		backoff *= 2
+		if backoff > cfg.MaxBackoff {
+			backoff = cfg.MaxBackoff
+		}
+	}
+
+	if lastErr == nil {
+		return fmt.Errorf("retry failed without error")
```
Comment (noa-lucent) [inline internal/retry/retry.go]: [major] **Unreachable "just-in-case" fallback — same issue from previous review, relocated.**

This was flagged as `[major]` in the previous round (on both `ensureRedis` and `subscribeWithRetry`). The code moved into the shared helper, but the pattern is unchanged.

When `cfg.MaxAttempts >= 1`, `lastErr` is always assigned on line 44 before the loop can exit without returning `nil`. The only way `lastErr` is still `nil` here is if `cfg.MaxAttempts <= 0` — which is a caller misconfiguration, not a normal runtime state.

Returning a synthetic `"retry failed without error"` string silently masks what would be a logic bug or an invalid config. Replace with a panic:

```go
if lastErr == nil {
    panic("retry.WithBackoff: exhausted retries with no recorded error")
}
```

Alternatively, validate `cfg.MaxAttempts >= 1` at the top of the function and return an error immediately for invalid config — that makes the contract explicit and eliminates this branch entirely:

```go
if cfg.MaxAttempts < 1 {
    return fmt.Errorf("retry: MaxAttempts must be >= 1, got %d", cfg.MaxAttempts)
}
```

File: internal/redis/subscriber.go [hunk 9]
```diff
@@ -125,6 +129,46 @@ func (s *Subscriber) Start(ctx context.Context) error {
...
+		)
+	}); err != nil {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		return nil, err
+	}
+
+	if pubsub == nil {
+		return nil, fmt.Errorf("redis subscribe failed")
```
Comment (noa-lucent) [inline internal/redis/subscriber.go]: [major] **Unreachable "just-in-case" fallback on `pubsub == nil`.**

If `WithBackoff` returns `nil` error, the operation closure succeeded — meaning `current` was non-nil (guarded at line 141) and `pubsub = current` executed (line 148). So `pubsub` is guaranteed non-nil when we reach line 165.

This is the same class of defensive fallback from the previous review, just in a new form. It masks an impossible state with a synthetic error instead of failing loudly.

Either remove the check entirely (trusting the contract of `WithBackoff`), or replace with a panic:

```go
if pubsub == nil {
    panic("subscribeWithRetry: WithBackoff succeeded but pubsub is nil")
}
```

File: internal/redis/subscriber_test.go (hunk 6 above)
Comment (noa-lucent) [reply internal/redis/subscriber_test.go:69]: This is still unaddressed. The subscriber test (`TestSubscriber_Start_ErrorOnSubscribe`) still only asserts `err != nil` without verifying that retries actually happened. Now that `subscribeWithRetry` delegates to the well-tested `retry.WithBackoff`, the urgency is lower — but adding a call counter to the mock would still catch regressions where the retry integration is accidentally bypassed. Leaving as [minor].

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Good progress — the shared `retry.WithBackoff` helper cleanly eliminates the duplication, and the new tests provide solid coverage of the retry mechanics. Four of seven previous threads are resolved.

Two major items remain:

1. **`retry.go:83` — Unreachable `lastErr == nil` fallback.** This is the same pattern from the first review, now consolidated in the helper. It still returns a synthetic error string for an impossible state instead of panicking or validating `MaxAttempts` upfront.

2. **`subscriber.go:165` — Unreachable `pubsub == nil` guard.** New in this revision. If `WithBackoff` returns nil error, `pubsub` is guaranteed non-nil by the closure logic. Same class of "just-in-case" defensive code.

One prior [minor] is also still open (subscriber test doesn't verify retry attempts), though its impact is reduced now that the retry helper itself is well-tested.

**Comment by casey-brooks** [comment]
## Summary
- validate retry config before executing backoff loop
- panic on impossible nil pubsub after successful retry

## Testing
- go build ./...
- go test ./...
  - passed: 4, failed: 0, skipped: 0
- go vet ./...
  - lint: no issues

**Review by noa-lucent — APPROVED** [review]
All prior feedback addressed. Clean changes:

- **`retry.go`**: Upfront `MaxAttempts >= 1` validation eliminates the unreachable fallback cleanly. The loop invariant is now explicit and the post-loop `return lastErr` is tight.
- **`subscriber.go`**: `pubsub == nil` replaced with a panic — correctly fails loud for an impossible state.

The shared `retry.WithBackoff` helper is well-structured, the tests cover the key paths (success-after-retries, exhaustion, context cancellation), and both call sites are concise.

Two optional follow-ups (non-blocking):
1. Consider adding a test for the `MaxAttempts < 1` validation path in `retry_test.go`.
2. The subscriber test could still benefit from a call counter on the mock to verify retry integration, though the helper's own tests reduce the risk.

LGTM — nice work iterating on this.
